Refactor classes

zerofivefiveseven opened the pull request:
wrote some methods to conversions.h
new structure
new conversions method
new base class
refactor calc methods

zerofivefiveseven committed:
Add SIFT implementation
Add class interface

zerofivefiveseven committed:
Add SIFT descriptor measure func

zerofivefiveseven committed:
Some codestyle fixes

zerofivefiveseven committed:
ADD the method of converting color to grayscale to class constructor
EDIT testcase. Now we read color image

zerofivefiveseven committed:
rename lib and associated header
codestyle fixes

zerofivefiveseven committed:
Add abstract class
create calc func instead of getkeypoints&getdescriptors
delete most of private fields and now them scope is just culc function

zerofivefiveseven committed:
Add abstract class
create calc func instead of getkeypoints&getdescriptors
delete most of private fields and now them scope is just culc function

zerofivefiveseven committed:
Merge remote-tracking branch 'origin/cpp/dev' into cpp/dev

# Conflicts:
#	src/feature_lib.cpp

TimPushkin commented on include/feature_lib.h:
```diff
@@ -0,0 +1,49 @@
+#include <opencv2/opencv.hpp>
+#include <opencv2/core/core.hpp>
+#include <opencv2/features2d.hpp>
+#include <opencv2/imgproc.hpp>
```
Put as much includes in the source file and leave here only those that are used in this header directly. You can find more information why [here](https://stackoverflow.com/questions/3002110/include-in-h-or-c-cpp).

zerofivefiveseven replied:
done

zerofivefiveseven replied:
read

TimPushkin commented on include/feature_lib.h:
```diff
@@ -0,0 +1,49 @@
+#include <opencv2/features2d.hpp>
+#include <opencv2/imgproc.hpp>
+
+namespace libstructs {
```
Let's name it `featurelib` as it contains the whole library

zerofivefiveseven replied:
done

TimPushkin commented on include/feature_lib.h:
```diff
@@ -0,0 +1,49 @@
+  float y;
+  float size;
+  float angle;
+  float response;
```
Let's name it `strength` as nobody known what does "response" mean. Different implementations name it differently anyway.

zerofivefiveseven replied:
done

TimPushkin commented on include/feature_lib.h:
```diff
@@ -0,0 +1,49 @@
+  float response;
+};
+
+struct CalcOutputStruct {
```
No point in naming a `struct` a Struct. Let's name it `DetectionResult`.

zerofivefiveseven replied:
done

TimPushkin commented on include/feature_lib.h:
```diff
@@ -0,0 +1,49 @@
+
+class FeatureDetector {
+ public:
+  // pure virtual function providing interface framework.
```
Seems too low level, let's remove this doc string.

zerofivefiveseven replied:
done

TimPushkin commented on include/feature_lib.h:
```diff
@@ -0,0 +1,49 @@
+class FeatureDetector {
+ public:
+  // pure virtual function providing interface framework.
+  virtual CalcOutputStruct calc(const std::vector<std::uint8_t> &input_vector) = 0;
```
Now I think, maybe "detect" would be better as the class is a detector? If you have some other thoughts, feel free to share as I'm not sure here.

TimPushkin replied:
And let's call the argument `image`

zerofivefiveseven replied:
I think that we calculate them, so 'calc' is what we need

TimPushkin replied:
But the process is actually called [feature detection](https://en.wikipedia.org/wiki/Feature_(computer_vision)), so we _detect_ them indeed

zerofivefiveseven replied:
done

TimPushkin commented on include/feature_lib.h:
```diff
@@ -0,0 +1,49 @@
+ public:
+  // pure virtual function providing interface framework.
+  virtual CalcOutputStruct calc(const std::vector<std::uint8_t> &input_vector) = 0;
+
```
Remove this blank line

zerofivefiveseven replied:
done

TimPushkin commented on include/feature_lib.h:
```diff
@@ -0,0 +1,49 @@
+ public:
+  SiftDetector(int width, int height);
+
+  CalcOutputStruct calc(const std::vector<std::uint8_t> &input_vector);
```
Clang-tidy suggests annotating it with `override`. Don't you have clang-tidy enabled in CLion?

zerofivefiveseven replied:
done
I just didn't pay attention
before i start work at this pr i notice that

TimPushkin commented on include/feature_lib.h:
```diff
@@ -0,0 +1,49 @@
+  int width;
+  int height;
+
+  cv::Ptr<cv::Feature2D> sift_instance_ = cv::SIFT::create();
```
I think, calling it just `sift_` would be shorter and still understandable

zerofivefiveseven replied:
done

TimPushkin commented on include/feature_lib.h:
```diff
@@ -0,0 +1,49 @@
+  cv::Ptr<cv::Feature2D> sift_instance_ = cv::SIFT::create();
+};
+
+}  //namespace libstructs
```
It would be better to add a space between `//` and `namespace` if the linter doesn't mind

zerofivefiveseven replied:
done

TimPushkin commented on src/feature_lib.cpp:
```diff
@@ -0,0 +1,23 @@
+#include "feature_lib.h"
+#include "conversions.h"
+namespace libstructs {
```
Add a blank line here

zerofivefiveseven replied:
done

TimPushkin commented on src/feature_lib.cpp:
```diff
@@ -0,0 +1,23 @@
+
+}  // namespace libstructs
+
+
```
Three blank lines is too much, a single one is enough :)

zerofivefiveseven replied:
done

TimPushkin commented on src/feature_lib.cpp:
```diff
@@ -0,0 +1,23 @@
+
+libstructs::CalcOutputStruct SiftDetector::calc(const std::vector<std::uint8_t> &input_vector) {
+  cv::Mat Image = VectorToMat(input_vector, height, width);
+  cv::cvtColor(Image, Image, cv::COLOR_BGR2GRAY);  //with that line test isn't working
```
I don't think this comment is still relevant

zerofivefiveseven replied:
done

TimPushkin commented on src/feature_lib.cpp:
```diff
@@ -0,0 +1,23 @@
+
+libstructs::CalcOutputStruct SiftDetector::calc(const std::vector<std::uint8_t> &input_vector) {
+  cv::Mat Image = VectorToMat(input_vector, height, width);
+  cv::cvtColor(Image, Image, cv::COLOR_BGR2GRAY);  //with that line test isn't working
```
We should probably do `RGB2GRAY` instead as it is much more popular nowadays

TimPushkin replied:
You may also need to tweak your test (by adding a `BGR2RGB` conversion) after this because OpenCV would read an image in BGR by default, I believe

zerofivefiveseven replied:
in work
deal with this tonight as a last priority

zerofivefiveseven replied:
nope, tomorrow

zerofivefiveseven replied:
i'll test it out
now RGB2GRAY flag

TimPushkin commented on src/feature_lib.cpp:
```diff
@@ -0,0 +1,23 @@
+  cv::cvtColor(Image, Image, cv::COLOR_BGR2GRAY);  //with that line test isn't working
+
+  std::vector<cv::KeyPoint> SiftKeyPoints;
+  cv::Mat SiftDescriptors;
```
These names don't match our code style

zerofivefiveseven replied:
done

TimPushkin commented on src/feature_lib.cpp:
```diff
@@ -0,0 +1,23 @@
+  std::vector<cv::KeyPoint> SiftKeyPoints;
+  cv::Mat SiftDescriptors;
+
+  sift_instance_->detectAndCompute(Image, cv::Mat(), SiftKeyPoints, SiftDescriptors);
```
Are you sure that passing an uninitialized `cv::Mat()` as a mask is ok? Shouldn't it be initialized to all 255 or something? It would be great if you could provide a link to the docs or an example.

zerofivefiveseven replied:
https://stackoverflow.com/questions/16365129/opencv-howto-use-mask-parameter-for-feature-point-detection-surf
I think it should be zero matrix

zerofivefiveseven replied:
with our image shape

zerofivefiveseven replied:
https://docs.opencv.org/3.4/d0/d13/classcv_1_1Feature2D.html
detect() method.
when i fix how image sizes kept in our class i ll deal with that

TimPushkin replied:
Seems like it should be a non-zero (255 probably) matrix sized the same as the image. See [here](https://answers.opencv.org/question/103087/behaviour-of-orb-keypoint-detection-with-masks/).

zerofivefiveseven replied:
done

TimPushkin commented on src/feature_lib.cpp:
```diff
@@ -0,0 +1,23 @@
+}
+
+libstructs::CalcOutputStruct SiftDetector::calc(const std::vector<std::uint8_t> &input_vector) {
+  cv::Mat Image = VectorToMat(input_vector, height, width);
```
This name doesn't match the code style

zerofivefiveseven replied:
done

TimPushkin commented on src/conversions.h:
```diff
@@ -0,0 +1,39 @@
+#include <cstdint>
+#include <opencv2/opencv.hpp>
+#include <opencv2/core/core.hpp>
+#include "feature_lib.h"
```
I believe, includes should go inside the include guards (i.e. after `ifndef` and `define`)

zerofivefiveseven replied:
done

TimPushkin commented on src/conversions.h:
```diff
@@ -0,0 +1,39 @@
+
+#ifndef FEATUREDETECT_SRC_CONVERSIONS_H_
+#define FEATUREDETECT_SRC_CONVERSIONS_H_
+cv::Mat VectorToMat(const std::vector<std::uint8_t> &inputVector, int Height, int Width) {
```
Add a blank line

zerofivefiveseven replied:
done

TimPushkin commented on src/conversions.h:
```diff
@@ -0,0 +1,39 @@
+
+#ifndef FEATUREDETECT_SRC_CONVERSIONS_H_
+#define FEATUREDETECT_SRC_CONVERSIONS_H_
+cv::Mat VectorToMat(const std::vector<std::uint8_t> &inputVector, int Height, int Width) {
```
This function name and the names of some arguments don't match the code style

TimPushkin replied:
Actually, all function names in this file don't, so fix them all, please

zerofivefiveseven replied:
>Actually, all function names in this file don't, so fix them all, please
public:
  int CountFooErrors(const std::vector<Foo>& foos) {
that example from google C++ codestyle link
VectorToMat matches, isn't it?

TimPushkin replied:
Yep, you are right here about the function names

TimPushkin replied:
Actually, let's make an exception here and name functions in `camelCase` with leading lowercase letter, because we intend to generate an interface for Java where all class members will have the same names as they do in C++.

zerofivefiveseven replied:
okay, should i rename whole project functions? or only in conversions file?

TimPushkin replied:
In fact, only the public ones will be ported, but let's rename all of them for the uniformity

TimPushkin commented on src/conversions.h:
```diff
@@ -0,0 +1,39 @@
+  return result;
+}
+
+std::vector<libstructs::KeyPoint> ConvertToStructure(const std::vector<cv::KeyPoint> &cv_key_points_vec) {
```
I think, naming the argument as `cv_key_points` would be enough

zerofivefiveseven replied:
done

TimPushkin commented on src/conversions.h:
```diff
@@ -0,0 +1,39 @@
+  std::vector<libstructs::KeyPoint> output;
+  for (auto &kp : cv_key_points_vec) {
+    libstructs::KeyPoint temp = {kp.pt.x, kp.pt.y, kp.angle, kp.response, kp.size};
+    output.reserve(5);
```
You can call `output.reserve(cv_key_points_vec.size())` once before the cycle -- this would be more efficient

TimPushkin replied:
And do you really need to allocate five elements to push a single one? I strongly believe, it takes the type and its size into account.

zerofivefiveseven replied:
done. i just I just didn't understand, now I understand

TimPushkin commented on src/conversions.h:
```diff
@@ -0,0 +1,39 @@
+  return output;
+}
+
+libstructs::CalcOutputStruct ConvertToOutputStructure(const std::vector<libstructs::KeyPoint> cv_key_points_vec,
```
These are our converted key points, not the ones from OpenCV. So, you can rename the parameter to `key_points`.

zerofivefiveseven replied:
done

TimPushkin commented on src/conversions.h:
```diff
@@ -0,0 +1,39 @@
+}
+
+libstructs::CalcOutputStruct ConvertToOutputStructure(const std::vector<libstructs::KeyPoint> cv_key_points_vec,
+                                                      const std::vector<uint8_t> descriptors) {
```
Clang-tidy suggests making both of these parameters references

zerofivefiveseven replied:
done

TimPushkin commented on src/conversions.h:
```diff
@@ -0,0 +1,39 @@
+  return output;
+}
+
+libstructs::CalcOutputStruct ConvertToOutputStructure(const std::vector<libstructs::KeyPoint> cv_key_points_vec,
```
Do we need this function at all? It can be replaced with `CalcOutputStruct {key_points, descriptors}`.

zerofivefiveseven replied:
done

TimPushkin commented on src/conversions.h:
```diff
@@ -0,0 +1,39 @@
+  output.descriptors = descriptors;
+  return output;
+}
+#endif //FEATUREDETECT_SRC_CONVERSIONS_H_
```
Add a blank line here

zerofivefiveseven replied:
done

TimPushkin commented on src/conversions.h:
```diff
@@ -0,0 +1,39 @@
+  output.descriptors = descriptors;
+  return output;
+}
+#endif //FEATUREDETECT_SRC_CONVERSIONS_H_
```
Add a space between `//` and `FEATUREDETECT_SRC_CONVERSIONS_H_`

zerofivefiveseven replied:
done

TimPushkin commented on examples/test.cpp:
```diff
@@ -1,6 +1,43 @@
+  return result;
+}
+
+int main(int argc, char *argv[]) {
```
Add some docs describing what a user has to pass as the CLI arguments

zerofivefiveseven replied:
do

zerofivefiveseven replied:
done

TimPushkin commented on examples/test.cpp:
```diff
@@ -1,6 +1,43 @@
+}
+
+int main(int argc, char *argv[]) {
+
```
Redundant blank line

zerofivefiveseven replied:
done

TimPushkin commented on examples/test.cpp:
```diff
@@ -1,6 +1,43 @@
+int main(int argc, char *argv[]) {
+
+  if (argc != 2) {
+    std::cerr << "Illegal CLI arguments num. Expected only filepath (including the command name) but was "
```
`std::cerr << "Illegal CLI arguments num. Expected only a file path but was " << argc << " arguments (including the command name)"` would be better

zerofivefiveseven replied:
done

TimPushkin commented on examples/test.cpp:
```diff
@@ -1,6 +1,43 @@
+
+  if (argc != 2) {
+    std::cerr << "Illegal CLI arguments num. Expected only filepath (including the command name) but was "
+              << argc;
```
Currently it can fit on a single line

zerofivefiveseven replied:
okay, done

TimPushkin commented on examples/test.cpp:
```diff
@@ -1,6 +1,43 @@
+    return 1;
+  }
+  std::string file_path = argv[1];
+  std::string image_path = cv::samples::findFile(file_path);
```
What do you need this line for?

zerofivefiveseven replied:
deleted. that line uses for "smart search" but as the result it wasn't smart at all

zerofivefiveseven replied:
done

TimPushkin commented on examples/test.cpp:
```diff
@@ -1,6 +1,43 @@
+
+  int width = img.cols;
+  int height = img.rows;
+  libstructs::SiftDetector sift_example(width, height);
```
Let's name it "detector"

TimPushkin replied:
Actually, you create an instance of this class again later

zerofivefiveseven replied:
done

TimPushkin commented on examples/test.cpp:
```diff
@@ -1,6 +1,43 @@
+  int height = img.rows;
+  libstructs::SiftDetector sift_example(width, height);
+
+  std::vector<uint8_t> test_vector = matToVector(img);
```
Let's name it "image_data"

zerofivefiveseven replied:
done

TimPushkin commented on examples/test.cpp:
```diff
@@ -1,6 +1,43 @@
+
+  std::vector<uint8_t> test_vector = matToVector(img);
+
+  libstructs::SiftDetector instance(width, height);
```
Why do you create a second instance of the detector here?

zerofivefiveseven replied:
remove that and use previous one

zerofivefiveseven replied:
done

TimPushkin commented on examples/test.cpp:
```diff
@@ -1,6 +1,43 @@
+  std::vector<uint8_t> test_vector = matToVector(img);
+
+  libstructs::SiftDetector instance(width, height);
+  libstructs::CalcOutputStruct test_output = instance.calc(test_vector);
```
Let's call it "output"

zerofivefiveseven replied:
done

TimPushkin replied:
Nope, it's still called `test_output`

zerofivefiveseven replied:
done

TimPushkin commented on examples/test.cpp:
```diff
@@ -1,6 +1,43 @@
+  libstructs::CalcOutputStruct test_output = instance.calc(test_vector);
+
+  std::cout << "Printing points" << std::endl;
+  for (auto &temp_kp : test_output.keypoints) {
```
"kp" instead of "temp_kp" would be shorter and still understandable

zerofivefiveseven replied:
done

TimPushkin commented on examples/test.cpp:
```diff
@@ -1,6 +1,43 @@
+  }
+
+  std::cout << "Printing descriptors" << std::endl;
+  for (auto &temp_uint : test_output.descriptors) {
```
Let's call it "desc_part". And, when you fix the descriptors array, print each descriptor on a separate line with parts saparated with spaces

TimPushkin commented on examples/test.cpp:
```diff
@@ -1,6 +1,43 @@
+
+  std::cout << "Printing points" << std::endl;
+  for (auto &temp_kp : test_output.keypoints) {
+    std::cout << temp_kp.x << "\t" << temp_kp.y << std::endl;
```
Let's print all parts of a key point including size, angle, etc.

zerofivefiveseven replied:
done

TimPushkin reviewed (changes requested):
Looks great overall, but descriptors should be represented by a `vector<vector<uint8_t>>` because each key point has an assiciated array of `uint8_t`s.

TimPushkin commented:
Also, from the architectural aspect every class and struct in `feature_lib.h` should've been placed into a separate file. But as far as I know, Scapix has problems with porting structs from different headers, so we have to place everything together.

TimPushkin commented on include/feature_lib.h:
```diff
@@ -0,0 +1,49 @@
+};
+
+class FeatureDetector {
+ public:
```
You should provide setters (and getters) for `width` and `height` so that a user won't have to reinitialize the whole class to change them

zerofivefiveseven replied:
done

zerofivefiveseven committed:
descriptor conversion method added
codestyle fixes
add getters setters methods
refactor constructor method
refactor class variables storing now they store in baseclass
edit example

TimPushkin commented on include/feature_lib.h:
```diff
@@ -0,0 +1,59 @@
+
+class FeatureDetector {
+ public:
+  virtual DetectionResult detect(const std::vector<std::uint8_t> &input_vector) = 0;
```
I think, this function can be made `const` too

zerofivefiveseven replied:
done

zerofivefiveseven replied:
renamed to vector_out

TimPushkin commented on include/feature_lib.h:
```diff
@@ -0,0 +1,59 @@
+
+class SiftDetector : public FeatureDetector {
+ public:
+  SiftDetector();
```
What do you need this constructor for? A default one is generated automatically. Though, I would suggest adding another constructor (and keep this one too) with two parameters: `width` and `height` -- set the according fields to the received values so that one won't have to call the corresponding setters.

TimPushkin replied:
After you add such constructor, you can simplify your test

zerofivefiveseven replied:
done

TimPushkin commented on include/feature_lib.h:
```diff
@@ -0,0 +1,59 @@
+  cv::Ptr<cv::Feature2D> sift_ = cv::SIFT::create();
+};
+
+}  // namespace featurelib
```
Remove a redundant space before `//` -- one is enough

zerofivefiveseven replied:
done

TimPushkin commented on src/CMakeLists.txt:
```diff
@@ -1,4 +1,4 @@
-set(SOURCE_FILES featurelib.cpp)
+set(SOURCE_FILES feature_lib.cpp conversions.h)
```
You don't need to put headers into sources

zerofivefiveseven replied:
done

TimPushkin commented on src/conversions.h:
```diff
@@ -0,0 +1,44 @@
+#include <cstdint>
+#include <opencv2/opencv.hpp>
+#include <opencv2/core/core.hpp>
+#include "feature_lib.h"
```
Some includes here are unused

zerofivefiveseven replied:
done

TimPushkin commented on src/conversions.h:
```diff
@@ -0,0 +1,44 @@
+  return {height, width, CV_8UC3, (std::uint8_t *) input_vector.data()};
+}
+
+std::vector<std::vector<uint8_t> > matTo2DVec(const cv::Mat &mat_in) {
```
Why the function above uses "vector" and this one uses "vec"? Choose one of them for consictency.

zerofivefiveseven replied:
renamed to vector_out

TimPushkin replied:
But why `vector_out`? I was talking about the function name, thought it would be just `matTo2DVector`.

zerofivefiveseven replied:
so i renamed func too

TimPushkin commented on src/conversions.h:
```diff
@@ -0,0 +1,44 @@
+  return {height, width, CV_8UC3, (std::uint8_t *) input_vector.data()};
+}
+
+std::vector<std::vector<uint8_t> > matTo2DVec(const cv::Mat &mat_in) {
```
Let's call the argument "mat" as it's shorter

zerofivefiveseven replied:
done

TimPushkin commented on src/conversions.h:
```diff
@@ -0,0 +1,44 @@
+      vec_out[i][j] = mat_in.at<uint8_t>(i, j);
+    }
+  }
+  return vec_out;
```
I think, you can do this conversion faster like [here](https://github.com/TimPushkin/SgbmDepthAndroid/blob/c2b5e96da5d2fbf49c605e4233c5c8f0318ef606/sgbmdepth/src/main/cpp/src/Conversions.h#L17) (I don't know why it is a `template` though as you need to define a concrete type -- change `float` to `uint8_t` in your case).

zerofivefiveseven replied:
done, I didn't delete the template, I decided that it would come later

zerofivefiveseven replied:
okay, it doesn't work, remove template

TimPushkin commented on src/conversions.h:
```diff
@@ -0,0 +1,44 @@
+  return vec_out;
+}
+
+std::vector<uint8_t> matTo1DVector(const cv::Mat &mat) {
```
Seems like this is not needed anymore

zerofivefiveseven replied:
done

TimPushkin commented on src/conversions.h:
```diff
@@ -0,0 +1,44 @@
+  return output;
+}
+
+#endif //  FEATUREDETECT_SRC_CONVERSIONS_H_
```
Remove a redundant space after `//` -- one is enough

zerofivefiveseven replied:
done

TimPushkin commented on src/feature_lib.cpp:
```diff
@@ -0,0 +1,21 @@
+
+namespace featurelib {
+
+SiftDetector::SiftDetector() {}
```
Clang-tidy suggests using `default` here

zerofivefiveseven replied:
done

TimPushkin commented on src/feature_lib.cpp:
```diff
@@ -0,0 +1,21 @@
+
+SiftDetector::SiftDetector() {}
+
+featurelib::DetectionResult SiftDetector::detect(const std::vector<std::uint8_t> &input_vector) {
```
Let's call the parameter "input" -- no need to repeat its type in its name

zerofivefiveseven replied:
done

TimPushkin commented on src/feature_lib.cpp:
```diff
@@ -0,0 +1,21 @@
+
+  std::vector<cv::KeyPoint> sift_key_points;
+  cv::Mat sift_descriptors;
+  cv::Mat mask(255 * cv::Mat::ones(height, width,CV_8U));
```
You can put it in class field so it won't be recreated for each image. And don't forget to update its sizes in width and height setters.

zerofivefiveseven replied:
done

TimPushkin commented on src/feature_lib.cpp:
```diff
@@ -0,0 +1,21 @@
+  cv::Mat sift_descriptors;
+  cv::Mat mask(255 * cv::Mat::ones(height, width,CV_8U));
+  sift_->detectAndCompute(image, mask, sift_key_points, sift_descriptors);
+  std::vector<KeyPoint> key_points_output = convertToStructure(sift_key_points);
```
Let's call it "key_points"

zerofivefiveseven replied:
done

TimPushkin replied:
Nope, still `key_points_output`. You acceentally renamed `sift_key_points` -- leave them as they were back then.

TimPushkin commented on src/feature_lib.cpp:
```diff
@@ -0,0 +1,21 @@
+  cv::Mat mask(255 * cv::Mat::ones(height, width,CV_8U));
+  sift_->detectAndCompute(image, mask, sift_key_points, sift_descriptors);
+  std::vector<KeyPoint> key_points_output = convertToStructure(sift_key_points);
+  std::vector<std::vector<uint8_t>> descriptors_output = matTo2DVec(sift_descriptors);
```
Let's call it "descriptors"

zerofivefiveseven replied:
done

TimPushkin replied:
Nope, still `descriptors_output`. You acceentally renamed `sift_descriptors` -- leave them as they were back then.

TimPushkin commented on src/feature_lib.cpp:
```diff
@@ -0,0 +1,21 @@
+  sift_->detectAndCompute(image, mask, sift_key_points, sift_descriptors);
+  std::vector<KeyPoint> key_points_output = convertToStructure(sift_key_points);
+  std::vector<std::vector<uint8_t>> descriptors_output = matTo2DVec(sift_descriptors);
+  return {key_points_output, descriptors_output};
```
Add some blank lines separating code blocks logically -- it's not easy to read seven different lines without any separation.

zerofivefiveseven replied:
done

TimPushkin commented on src/feature_lib.cpp:
```diff
@@ -0,0 +1,21 @@
+  return {key_points_output, descriptors_output};
+}
+
+}  // namespace featurelib
```
Remove the redundant space before `//` -- one is enough

zerofivefiveseven replied:
done

TimPushkin commented on examples/test.cpp:
```diff
@@ -1,6 +1,43 @@
+#include "../include/feature_lib.h"
+#include <iostream>
+#include <opencv2/core.hpp>
+#include <opencv2/imgcodecs.hpp>
```
Rearrange includes as the code style [suggests](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes)

zerofivefiveseven replied:
done

TimPushkin commented on examples/test.cpp:
```diff
@@ -1,6 +1,43 @@
+  cont.reshape(1, 1).copyTo(result);  // copy from a 1D Mat
+  return result;
+}
+//Please inter only filepath to command line arguments
```
Add a blank line between `matToVector` and this doc

zerofivefiveseven replied:
done

TimPushkin commented on examples/test.cpp:
```diff
@@ -1,6 +1,43 @@
+  cont.reshape(1, 1).copyTo(result);  // copy from a 1D Mat
+  return result;
+}
+//Please inter only filepath to command line arguments
```
1. Add a space between `//` and the comment itself.
2. Fix typos: it's "enter" not "inter".
3. You don't write docs like this: no need to write "please" -- you describe what the function does, what args it accepts, and what it returns. And you make it as short as possible. You can see some examples [in the code style](https://google.github.io/styleguide/cppguide.html#Function_Comments).

zerofivefiveseven replied:
done

TimPushkin commented on examples/test.cpp:
```diff
@@ -1,6 +1,43 @@
+//Please inter only filepath to command line arguments
+int main(int argc, char *argv[]) {
+  if (argc != 2) {
+    std::cerr << "Illegal CLI arguments num. Expected only a file path but was " << argc << " arguments (including the command name)"<<std::endl;
```
Now this line exceeds the length limit. Press `Ctrl+Alt+L` and CLion will fix this for you.

zerofivefiveseven replied:
done

TimPushkin commented on examples/test.cpp:
```diff
@@ -1,6 +1,43 @@
+
+  featurelib::SiftDetector detector;
+  detector.setWidth(width);
+  detector.setHeight(height);
```
This setters will be gone after you add a new constructor to the detector class

zerofivefiveseven replied:
done

TimPushkin commented on examples/test.cpp:
```diff
@@ -1,6 +1,43 @@
+  {
+    for ( const auto &s : row ) std::cout << s << ' ';
+    std::cout << std::endl;
+  }
```
Wrong formatting here, press `Ctrl+Alt+L` to fix this

zerofivefiveseven replied:
done

TimPushkin commented on examples/test.cpp:
```diff
@@ -1,6 +1,43 @@
+  std::cout << "Printing descriptors" << std::endl;
+  for ( const auto &row : test_output.descriptors )
+  {
+    for ( const auto &s : row ) std::cout << s << ' ';
```
What does `s` mean? Pick a name that makes more sense.

zerofivefiveseven replied:
done

TimPushkin reviewed (changes requested):
1. Use `auto` where possible if it doesn't make the code less clear. You can read more about when to use `auto` [here](https://google.github.io/styleguide/cppguide.html#Type_deduction).
2. Let's rename `test.cpp` to `detection_example.cpp`.

zerofivefiveseven committed:
new constructor
remove setters
refactor matto2dvec and renamed it
edit codestyle & docstrings
remove
refactor mask storing

TimPushkin commented on examples/detection_example.cpp:
```diff
@@ -0,0 +1,45 @@
+}
+
+// Calculates key points and image descriptors
+// Takes a filepath as input
```
"image file path" would be better

zerofivefiveseven replied:
done

TimPushkin commented on examples/detection_example.cpp:
```diff
@@ -0,0 +1,45 @@
+
+// Calculates key points and image descriptors
+// Takes a filepath as input
+// Returns data in the structure declared in associated header
```
It actually returns nothing, it prints keypoints and descriptors to `stdout`

zerofivefiveseven replied:
done

zerofivefiveseven committed:
edit codestyle

TimPushkin commented on include/feature_lib.h:
```diff
@@ -0,0 +1,53 @@
+ protected:
+  int width;
+  int height;
+  cv::Mat mask = cv::Mat(255 * cv::Mat::ones(height, width, CV_8U));
```
Bring back the setters. We want to allow our users to change these without recreating an object. And don't forget to recalculate the mask in these setters.

zerofivefiveseven replied:
done!

TimPushkin commented on include/feature_lib.h:
```diff
@@ -0,0 +1,53 @@
+class SiftDetector : public FeatureDetector {
+ public:
+  SiftDetector();
+  SiftDetector(int width, int height);
```
Add a blank line to separate the constructors

zerofivefiveseven replied:
done

TimPushkin commented on include/feature_lib.h:
```diff
@@ -0,0 +1,53 @@
+  SiftDetector();
+  SiftDetector(int width, int height);
+
+  DetectionResult detect(const std::vector<std::uint8_t> &input) const;
```
Bring back the `override` specifier. Clang-tidy suggests it -- pay attention to its warnings.

zerofivefiveseven replied:
done

TimPushkin commented on src/conversions.h:
```diff
@@ -0,0 +1,36 @@
+#ifndef FEATUREDETECT_SRC_CONVERSIONS_H_
+#define FEATUREDETECT_SRC_CONVERSIONS_H_
+
+#include <opencv2/opencv.hpp>
```
I think, you need `opencv2/core.hpp` here

zerofivefiveseven replied:
it still building without, but i ll do that

TimPushkin replied:
You can remove `#include <opencv2/opencv.hpp>` here. I meant that you should have been replaced one with another.

zerofivefiveseven replied:
oh i am so sorry

zerofivefiveseven replied:
done

TimPushkin commented on src/conversions.h:
```diff
@@ -0,0 +1,36 @@
+  auto rowLen = cont.cols * cont.channels();
+  for (auto i = 0; i < cont.rows; i++) {
+    const auto *rowPtr = cont.ptr<uint8_t>(i);
+    result.push_back(std::vector<uint8_t>(rowPtr, rowPtr + rowLen));
```
Clang-tidy suggests using `emplace_back` here

zerofivefiveseven replied:
done

TimPushkin commented on src/feature_lib.cpp:
```diff
@@ -0,0 +1,28 @@
+#include "feature_lib.h"
+#include "conversions.h"
```
You need to include `opencv2/imgproc.hpp` -- currently you are getting `cv::COLOR_RGB2GRAY` through a transitive dependency from `conversions.h` which is bad

zerofivefiveseven replied:
done

TimPushkin reviewed (changes requested):
You can still add `auto` in a lot of places

zerofivefiveseven committed:
new setters which update sizes of mask matrix
edit codestyle

TimPushkin commented on examples/detection_example.cpp:
```diff
@@ -0,0 +1,45 @@
+
+// Calculates keypoints and image descriptors
+// Takes an image file path as input
+// prints key points and image descriptors to stdout declared in associated header
```
1. Start this sentence with a capital letter.
2. `to stdout declared in associated header` -- what? You need to remove the info about the header, I believe.

zerofivefiveseven replied:
my bad

TimPushkin commented on include/feature_lib.h:
```diff
@@ -0,0 +1,62 @@
+
+  void setWidth(int width) {
+    mask.reshape(1, std::vector<int>{height, width});
+  }
```
You must also update the corresponding fields in both of these setters -- otherwise what's the point?

TimPushkin commented on include/feature_lib.h:
```diff
@@ -0,0 +1,62 @@
+  }
+
+  void setWidth(int width) {
+    mask.reshape(1, std::vector<int>{height, width});
```
I think, you cannot use `reshape` here (and for the height too) as the number of elements changes. You need to create a new matrix or change the size of this one somehow.

zerofivefiveseven replied:
for example, when only the width increases and the height remains the same, you need to somehow add columns, so I will recreate it

TimPushkin replied:
Yep. There is also `cv::resize`, but I think that recreating would be faster because this resizing involves interpolation

TimPushkin commented on include/feature_lib.h:
```diff
@@ -0,0 +1,62 @@
+
+ protected:
+  int width;
+  int height;
```
You need to initialize them to something, or they'll stay uninitialized in case of using the default constructor. CLion gives you a warning -- do you even read them?

zerofivefiveseven replied:
Yep, this is because the names are the same as those in the setters
even i initialize them to something clion is still giving me warnings

zerofivefiveseven replied:
when i change the parameter in the setter the warning goes away

zerofivefiveseven replied:
maybe height_input and width_input ass parameters and also i initialize protected fields&

TimPushkin replied:
Yep, the warning I'm talking about appears in the default constructor's
 implementation.

Let the names be just `value` in the setters: `void setWidth(int value)`

TimPushkin commented on src/feature_lib.cpp:
```diff
@@ -0,0 +1,29 @@
+#include "feature_lib.h"
+#include "conversions.h"
+#include <opencv2/imgproc.hpp>
```
Fix the order according to the code style

zerofivefiveseven replied:
![image](https://user-images.githubusercontent.com/33483629/180479652-809083bc-4095-4862-8666-1fde4b34acb1.png)
https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

zerofivefiveseven replied:
That's better?

TimPushkin replied:
Yep, but remove that blank line

TimPushkin reviewed (changes requested):
You got the setters wrong again :(

zerofivefiveseven committed:
edit codestyle
edit setters again
edit docstring
initialize width and height in header

TimPushkin commented on src/feature_lib.cpp:
```diff
@@ -0,0 +1,29 @@
+
+featurelib::DetectionResult SiftDetector::detect(const std::vector<std::uint8_t> &input) const {
+  auto image = vectorToMat(input, height, width);
+  cv::cvtColor(image, image, cv::COLOR_RGB2GRAY);//
```
Remove `//` in the end

zerofivefiveseven replied:
ok

TimPushkin reviewed (changes requested):
Other than this small detail, looks good to me

achains commented on examples/detection_example.cpp:
```diff
@@ -0,0 +1,45 @@
+#include <iostream>
+#include <opencv2/core.hpp>
+#include <opencv2/imgcodecs.hpp>
+#include "../include/feature_lib.h"
```
Is relative import really necessary here, since we have linked feature library to this target?

zerofivefiveseven replied:
Nope, this works without a relative path as well. And the Google cpp codestyle says to avoid Unix ../ characters. I rewrite it like #include<"file.h">

achains reviewed (commented):
Nice work, but it'd be better to remove relative import in examples

anastasiia-kornilova commented on include/feature_lib.h:
```diff
@@ -0,0 +1,64 @@
+ public:
+  virtual DetectionResult detect(const std::vector<std::uint8_t> &input_vector) const = 0;
+
+  void setHeight(int value) {
```
Подскажите по какому гайду мы живем, что методы определяются в хедере?

zerofivefiveseven replied:
moved all base class definitions to .cpp file
declarations of all of this methods still in .h file

anastasiia-kornilova commented on include/feature_lib.h:
```diff
@@ -0,0 +1,64 @@
+  cv::Mat mask = cv::Mat(255 * cv::Mat::ones(height, width, CV_8U));
+};
+
+class SiftDetector : public FeatureDetector {
```
Один класс на один файл или это тоже можно в нашем стайл гайде?

anastasiia-kornilova replied:
Обсудили с Тимофеем -- так нужно скапиксу, пока отправляется в issues

anastasiia-kornilova commented on include/feature_lib.h:
```diff
@@ -0,0 +1,64 @@
+#ifndef FEATUREDETECT_FEATURELIB_H_
```
Лучше обзывать всё Keypoints -- features слишком многозначное слово, можно пока занести в issue и поправить на след итерации

anastasiia-kornilova commented on include/feature_lib.h:
```diff
@@ -0,0 +1,64 @@
+ protected:
+  int width = 0;
+  int height = 0;
+  cv::Mat mask = cv::Mat(255 * cv::Mat::ones(height, width, CV_8U));
```
Мне неочевидно зачем нужна маска, давайте оставим коммент.

zerofivefiveseven replied:
![image](https://user-images.githubusercontent.com/33483629/180544484-6581a7b8-3e8e-4a27-8b2a-c542b7a42b18.png)
made a comment in docstring style
should I use a multi-line comment on the left?

anastasiia-kornilova commented on include/feature_lib.h:
```diff
@@ -0,0 +1,64 @@
+
+  void setHeight(int value) {
+    height = value;
+    mask = cv::Mat(255 * cv::Mat::ones(height, width, CV_8U));
```
Кажется этот конструктор здесь уместнее -- https://docs.opencv.org/4.x/d3/d63/classcv_1_1Mat.html#a3620c370690b5ca4d40c767be6fb4ceb

zerofivefiveseven replied:
Done, should i use Initializer List there too?

anastasiia-kornilova replied:
Not clear for me how do you plan to use it here, could you provide example?

zerofivefiveseven replied:
my bad, i messed up.. again

anastasiia-kornilova commented on src/conversions.h:
```diff
@@ -0,0 +1,36 @@
+#include "feature_lib.h"
+
+cv::Mat vectorToMat(const std::vector<std::uint8_t> &input_vector, int height, int width) {
+  return {height, width, CV_8UC3, (std::uint8_t *) input_vector.data()};
```
Лучше использовать явные касты, в данном случае static_cast. Почитать почему он умнее обычного explicit cast можно здесь https://cplusplus.com/doc/oldtutorial/typecasting/ .

zerofivefiveseven replied:
Static_cast from 'const unsigned char *' to 'std::uint8_t *' (aka 'unsigned char *') is not allowed 
Maybe const_cast instead?
const_cast
This type of casting manipulates the constness of an object, either to be set or to be removed. For example, in order to pass a const argument to a function that expects a non-constant parameter
if i use that type of casting it works

anastasiia-kornilova replied:
Yep, you are right

anastasiia-kornilova commented on src/conversions.h:
```diff
@@ -0,0 +1,36 @@
+  std::vector<featurelib::KeyPoint> output;
+  output.reserve(key_points.size());
+  for (auto &kp : key_points) {
+    featurelib::KeyPoint temp = {kp.pt.x, kp.pt.y, kp.angle, kp.response, kp.size};
```
Ужаснее названия `temp` может быть только `a`, почему нельзя это в одну строчку сделать без временных переменных?

zerofivefiveseven replied:
Rewrote the method without using temporary variable
![image](https://user-images.githubusercontent.com/33483629/180515595-e2fcb4b8-f389-4196-b727-7d6a7e393c85.png)
should i use some cast things there? i ll read link above rn

anastasiia-kornilova replied:
What would you like to cast to what? It is already the same type that is stored in vector.

zerofivefiveseven replied:
my bad, i messed up..

anastasiia-kornilova commented on src/feature_lib.cpp:
```diff
@@ -0,0 +1,29 @@
+SiftDetector::SiftDetector() = default;
+
+SiftDetector::SiftDetector(int width, int height) {
+  this->width = width;
```
Почитайте про то, что такое Initializer List in C++  и почему лучше так инициализировать параметры чем как у Вас

zerofivefiveseven replied:
done.
I initialized the base class using this technique and in the initialization of the child class used that too
![image](https://user-images.githubusercontent.com/33483629/180574828-6d4260a9-d22b-4a8a-b02c-f405f23970cb.png)
![image](https://user-images.githubusercontent.com/33483629/180574861-bac0de72-f56b-4894-a263-397c1d57a74a.png)

TimPushkin commented on BuildOpenCV.cmake:
```diff
@@ -26,7 +26,14 @@ if (NOT EXISTS ${OPENCV_DOWNLOAD_DIR}/opencv-${OPENCV_VERSION}/CMakeLists.txt)
+        message(STATUS "Downloading OpenCV-extra")
+        file(
+                DOWNLOAD
+                https://github.com/opencv/opencv_contrib/archive/refs/heads/4.x.zip
```
Donwload the same version as the version of the main package. You can do this from tags, for example: `https://github.com/opencv/opencv_contrib/archive/refs/tags/${OPENCV_VERSION}.zip`

TimPushkin replied:
Closing because this and all the conversions below as this change has been removed

TimPushkin commented on BuildOpenCV.cmake:
```diff
@@ -26,7 +26,14 @@ if (NOT EXISTS ${OPENCV_DOWNLOAD_DIR}/opencv-${OPENCV_VERSION}/CMakeLists.txt)
                 ${OPENCV_DOWNLOAD_DIR}/${OPENCV_VERSION}.zip
```
Change it to `${OPENCV_DOWNLOAD_DIR}/opencv-${OPENCV_VERSION}.zip`. And don't forget to also change the `NOT EXISTS` check.

TimPushkin commented on BuildOpenCV.cmake:
```diff
@@ -26,7 +26,14 @@ if (NOT EXISTS ${OPENCV_DOWNLOAD_DIR}/opencv-${OPENCV_VERSION}/CMakeLists.txt)
+        file(
+                DOWNLOAD
+                https://github.com/opencv/opencv_contrib/archive/refs/heads/4.x.zip
+                ${OPENCV_DOWNLOAD_DIR}/opencv-extra.zip
```
Change it to `${OPENCV_DOWNLOAD_DIR}/opencv-extra-${OPENCV_VERSION}.zip`

TimPushkin commented on BuildOpenCV.cmake:
```diff
@@ -26,7 +26,14 @@ if (NOT EXISTS ${OPENCV_DOWNLOAD_DIR}/opencv-${OPENCV_VERSION}/CMakeLists.txt)
                 ${OPENCV_DOWNLOAD_DIR}/${OPENCV_VERSION}.zip
                 SHOW_PROGRESS
         )
-        message(STATUS "Downloading OpenCV - done")
```
Bring this back, please. And add a blank line after it.

TimPushkin commented on BuildOpenCV.cmake:
```diff
@@ -26,7 +26,14 @@ if (NOT EXISTS ${OPENCV_DOWNLOAD_DIR}/opencv-${OPENCV_VERSION}/CMakeLists.txt)
                 SHOW_PROGRESS
         )
-        message(STATUS "Downloading OpenCV - done")
+        message(STATUS "Downloading OpenCV-extra")
```
Add a separate check here that this hasn't yet been downloaded

TimPushkin commented on BuildOpenCV.cmake:
```diff
@@ -36,6 +43,14 @@ if (NOT EXISTS ${OPENCV_DOWNLOAD_DIR}/opencv-${OPENCV_VERSION}/CMakeLists.txt)
     )
     message(STATUS "Extracting OpenCV - done")
+
+    message(STATUS "Extracting OpenCV-extra")
```
Add a separate check that this hasn't yet been extracted (like the one I made for the standard OpenCV)

TimPushkin commented on BuildOpenCV.cmake:
```diff
@@ -65,6 +80,7 @@ set(
         OPENCV_CMAKE_ARGS  # https://docs.opencv.org/4.6.0/db/d05/tutorial_config_reference.html
         # General
         -DBUILD_LIST=${OPENCV_MODULES}
+        -DOPENCV_EXTRA_MODULES_PATH=${OPENCV_DOWNLOAD_DIR}/opencv-extra/opencv_contrib-4.x/modules
```
Modify this path according to the changes I described above

TimPushkin replied:
Also, add a `TODO` to specify only the extra modules we need. I believe, we only need `xfeatures2d`.

TimPushkin reviewed (changes requested):
Better move the `opencv-contrib` downloading and extracting into a separate block with the same check that the standard `opencv` has. Maybe even create a function for it because the same code is used twice.

anastasiia-kornilova commented on include/feature_lib.h:
```diff
@@ -0,0 +1,56 @@
+  int height = 0;
+  // You can limit your search to a set of regions with a mask.
+  // It must be a 8-bit integer matrix with non-zero values in the region of interest.
+  cv::Mat mask = cv::Mat(255 * cv::Mat::ones(height, width, CV_8U));
```
Could you change here also to more suitable constructor as we discussed in other files?

zerofivefiveseven commented:
I did something with commits, and somehow strangely rolled back the changes in BuildOpenCV.cmake

TimPushkin commented:
You did something wrong: you pushed the same commits and two lines related to the `opencv-contrib` changes are still there

zerofivefiveseven commented:
rolled back changes in BuildOpenCV.cmake

zerofivefiveseven committed:
change cast type in conversions.h
    change mat constructor in conversions.h
    edit constructor of base class now it use initializer list constructor
    remove default constructors
    now child class use initializer list constructor
    move definitions of base class to .cpp

zerofivefiveseven committed:
change Mat constructor of a base class field named mask

zerofivefiveseven commented:
rewrite history 
add last requested correction
